java/ql/src/experimental/Security/CWE/CWE-299/RevocationCheckingLib.qll

```diff
@@ -0,0 +1,181 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking2
+import DataFlow
+
+/**
+ * A taint-tracking configuration for disabling revocation checking.
+ */
+class DisabledRevocationCheckingConfig extends TaintTracking::Configuration {
+  DisabledRevocationCheckingConfig() { this = "DisabledRevocationCheckingConfig" }
+
+  override predicate isSource(DataFlow::Node source) {
+    exists(BooleanLiteral b | b.getBooleanValue() = false | source.asExpr() = b)
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof SetRevocationEnabledSink }
+}
+
+/**
+ * A sink that disables revocation checking,
+ * i.e. calling `PKIXParameters.setRevocationEnabled(false)`
+ * without setting a custom revocation checker in `PKIXParameters`.
+ */
+class SetRevocationEnabledSink extends DataFlow::ExprNode {
+  SetRevocationEnabledSink() {
+    exists(MethodAccess setRevocationEnabledCall |
+      setRevocationEnabledCall.getMethod() instanceof SetRevocationEnabledMethod and
+      setRevocationEnabledCall.getArgument(0) = getExpr() and
+      not exists(
+        SettingRevocationCheckerConfig config, DataFlow2::PathNode source, DataFlow2::PathNode sink
+      |
+        config.hasFlowPath(source, sink) and
+        sink.getNode().(SettingRevocationCheckerSink).getVariable() =
+          setRevocationEnabledCall.getQualifier().(VarAccess).getVariable()
+      )
+    )
+  }
+}
+
+/**
+ * A dataflow config for tracking a custom revocation checker.
+ */
+class SettingRevocationCheckerConfig extends DataFlow2::Configuration {
+  SettingRevocationCheckerConfig() {
+    this = "DisabledRevocationChecking::SettingRevocationCheckerConfig"
+  }
+
+  override predicate isSource(DataFlow::Node source) {
+    source instanceof GetRevocationCheckerSource
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof SettingRevocationCheckerSink }
+
+  override predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
```

Comment:
Do you want to also cover the [`List.of` methods](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/List.html#of(E)) (except `of()` which returns an empty list) added in Java 9?

Reply:
Yeah, makes sense.